src/main/java/org/sonarlint/intellij/ui/AbstractFindingsPanel.java

```diff
@@ -0,0 +1,200 @@
+import org.sonarlint.intellij.common.util.SonarLintUtils;
+import org.sonarlint.intellij.editor.EditorDecorator;
+import org.sonarlint.intellij.ui.nodes.IssueNode;
+import org.sonarlint.intellij.ui.nodes.LiveSecurityHotspotNode;
+import org.sonarlint.intellij.ui.tree.FlowsTree;
+import org.sonarlint.intellij.ui.tree.FlowsTreeModelBuilder;
+import org.sonarlint.intellij.ui.tree.IssueTree;
+import org.sonarlint.intellij.ui.tree.IssueTreeModelBuilder;
+import org.sonarlint.intellij.ui.tree.SecurityHotspotTree;
+import org.sonarlint.intellij.ui.tree.SecurityHotspotTreeModelBuilder;
+
+public abstract class AbstractFindingsPanel extends SimpleToolWindowPanel {
+  private static final String ID = "SonarLint";
+  private static final int RULE_TAB_INDEX = 0;
+  private static final int LOCATIONS_TAB_INDEX = 1;
+  protected final Project project;
+  protected SonarLintRulePanel rulePanel;
+  protected JBTabbedPane detailsTab;
+  protected Tree tree;
+  protected IssueTreeModelBuilder treeBuilder;
+  protected FlowsTree flowsTree;
+  protected FlowsTreeModelBuilder flowsTreeBuilder;
+  private ActionToolbar mainToolbar;
+  protected SecurityHotspotTreeModelBuilder shTreeBuilder;
+  protected Tree shTree;
+
+  protected AbstractFindingsPanel(Project project) {
+    super(false, true);
+    this.project = project;
+
+    createFlowsTree();
+    createIssuesTree();
+    createSecurityHotspotsTree();
+    createTabs();
+    handleListener();
+    disableSecurityHotspotTree();
+  }
+
+  public void refreshToolbar() {
+    mainToolbar.updateActionsImmediately();
+  }
+
+  public void enableHotspotTree() {
+    tree.setShowsRootHandles(true);
+    shTree.setShowsRootHandles(true);
+    shTree.setVisible(true);
+  }
+
+  public void disableSecurityHotspotTree() {
+    tree.setShowsRootHandles(false);
+    shTree.setShowsRootHandles(false);
+    shTree.setVisible(false);
+  }
+
+  private void handleListener() {
+    shTree.addTreeSelectionListener(this::findingTreeSelectionChanged);
+    tree.addTreeSelectionListener(this::findingTreeSelectionChanged);
+  }
+
+  private void createSecurityHotspotsTree() {
+    shTreeBuilder = new SecurityHotspotTreeModelBuilder();
+    var model = shTreeBuilder.createModel();
+    shTree = new SecurityHotspotTree(project, model);
+    manageInteraction(shTree);
+  }
+
+  private void createTabs() {
+    // Flows panel with tree
+    var flowsPanel = ScrollPaneFactory.createScrollPane(flowsTree, true);
+    flowsPanel.getVerticalScrollBar().setUnitIncrement(10);
+
+    // Rule panel
+    rulePanel = new SonarLintRulePanel(project);
+
+    detailsTab = new JBTabbedPane();
+    detailsTab.insertTab("Rule", null, rulePanel, "Details about the rule", RULE_TAB_INDEX);
+    detailsTab.insertTab("Locations", null, flowsPanel, "All locations involved in the finding", LOCATIONS_TAB_INDEX);
+  }
+
+  protected void findingTreeSelectionChanged(TreeSelectionEvent e) {
+    if (e.getSource() instanceof IssueTree) {
```

Comment:
You could register 2 different listeners instead of using instanceof. Or as the 2 branches share code in common maybe we could factorize using LiveFinding? Not sure about that

Reply:
I factorized some code and split in 2 listeners